victorjambo/WeConnect
# Implement database with PG

**What does this PR do?**
Implements the logic of the endpoints with Postgres and SQLAlchemy

**Description of Task to be completed?**

The following are functions the user can perform given different endpoints

User can register and Login
User gets an email after registration to activate the account
User can request reset password and sent via mail (forgot password)
User can register a new business
User can view all businesses
User can search for businesses
User can filter by location or category
User can review a business
User can update or delete a business or review if authorized
User can reset password

How should this be tested?
Clone this repo

```bash
cd WeConnect
```

Install the packages if you haven't.
```bash
$ pip install -r requirements.txt
```

First, create a Postgres database. Then set the URI as an environment
variable. See below how to set environment variables
Once db is set just run;
```bash
flask db upgrade
```

Tests these endpoints with Postman

**EndPoint** | **Functionality**
--- | ---
POST `/api/v2/auth/register` | Creates a user account 
POST `/api/v2/auth/login` | Logs in a user
POST `/api/v2/auth/logout` | Logs out a user
POST `/api/v2/auth/reset-password` | Password reset
POST  `/api/v2/businesses` | Register a business
PUT `/api/v2/business/<businessId>` | Updates a business profile
DELETE `/api/v2/businesses/<businessId>` | Remove a business
GET  `/api/v2/businesses` | Retrieves all businesses
GET  `/api/v2/businesses/<businessId>` | Get a business 
POST  `/api/v2/businesses/<businessId>/reviews` | Add a review for a business
GET  `/api/v2/businesses/<businessId>/reviews` | Get all reviews for a business
GET  `/api/v2/notifications` | Get all notifications for current logged in user

*miscellaneous endpoint*

**EndPoint** | **Functionality**
--- | ---
GET `/api/v2/verify?key=<hash_key>&name=<username>` | activates user email
GET `/api/v2/businesses/?q=<query>` | search business name
GET `/api/v2/businesses/?q=<query>&location=<location>` | search in the specified location
GET `/api/v2/businesses/?q=<query>&category=<category>` | search in the specified category
GET `/api/v2/businesses/?location=<location>&category=<category>` | filter businesses
GET `/api/v2/businesses/?page=2&limit=10` | sets pagination and number of results per page

Run app with;

```bash
$ python3 app.py
```

To run your test cases use

```bash
$ nosetests
```

**Any background context you want to provide?**
Built with python3

*Environment variables*
Set environment variables for `SECRET`, `ENVIRON` `DATABASE_URL`, `FLASK_APP`, `GMAIL_MAIL` and `GMAIL_PASSWORD`
> `SECRET` is your secret key for the app

> `ENVIRON` is the enviroment you are running on. Should be either `Production`, `Development` or `Testing`. For this PR you can use `Development` NOTE: its case sensitive

> `GMAIL_MAIL` and `GMAIL_PASSWORD` go to `https://myaccount.google.com/apppasswords` and generate app password for your API. For Development, you can use dummy email and password since the mail is suppressed

> `DATABASE_URL` this is your Postgres database URI. Ensure you have created a postgres database first.

> `FLASK_APP` value should be `app.py`. That is the file where our app starts from. `FLASK_APP=app.py`
